firedrake/slate/slate.py

```diff
@@ -209,10 +219,32 @@ def coeff_map(self):
         to the split global coefficient numbers.
         The split coefficients are defined on the pieces of the originally mixed function spaces.
         """
-        return tuple((n, tuple(range(len(c.split()))))
-                     if isinstance(c, Function) or isinstance(c, Constant)
-                     else (n, (0,))
-                     for n, c in enumerate(self.coefficients()))
+        coeff_map = []
+        orig_to_pos_dict = {}
+        for c in self.coefficients():
+            m = len(coeff_map)
+            # BlockFunctions for c need special treatment because:
+            # 1) There can be multiple BlockFunctions defined on the same original function
+            # -> The split functions in c might need to be appended to an already exisiting entry in the map
+            # 2) BlockAssembledVectors can appear in any order they like in the Slate expression
+            # -> The split functions need to be appended in the same order into the map
+            if isinstance(c, BlockFunction):  # for block assembled vectors
+                # Did we already add a part of the originial function of this BlockFunction to the map?
+                orig_to_pos = tuple(*filter(lambda item: item[0] == c.orig_function, orig_to_pos_dict.items()))
+                pos = orig_to_pos[1] if orig_to_pos else m
+                split_map = c.indices[0] if pos == m else coeff_map[pos][1]+c.indices[0]
```

Comment:
The following example fails:
```
from firedrake import *
  

mesh = UnitSquareMesh(1, 1)
V0 = FunctionSpace(mesh, "DG", 0)
V1 = FunctionSpace(mesh, "DG", 1)
V = V0 * V1
u = TrialFunction(V)
v = TestFunction(V)
K = Tensor(inner(u, v) * dx)
f = AssembledVector(assemble(inner(as_vector([1., 1.]), v)*dx))
expr = (K*f).blocks[0] + f.blocks[0]
assemble(expr, form_compiler_parameters={"slate_compiler": {"optimise": True}})
```
with:
```
AssertionError: KernelInfo must be generated with a coefficient map that maps EXACTLY all cofficients there are in its arguments attribute.
```
I think this is because we use tuple-add here instead of set-update:

`coeff_map` for this example is indeed:
```
((0, (0, 1, 0)),)
```

Reply:
If I use a plain set we loose the order on the split_map, but I will push a fix with a sorted set when I get to it. Thanks!

Reply:
Does the order in split_map need to be preserved merely to be consistent with the order of `kernel_arg`? If so, I think the order can be arbitrary once you amend `collect_coefficients` to use `coeff_map`'s order (see below).

Reply:
I pushed a fix for this here https://github.com/firedrakeproject/firedrake/pull/2446/commits/92c705728a5ad5953f0a14ea44d50b569a2d15d9. As I said in a comment below I think that coefficients map does not carry enough information to the collect the coefficients for the kernel argument generation, it's just indices and nothing else

Reply:
If we flip the order, the above example still fails:

From:
```
expr = (K*f).blocks[0] + f.blocks[0]
```
To:
```
expr = f.blocks[0] + (K*f).blocks[0]
```
You could directly fix this, but see below comment.